File: tyhp-lang/vscode/src/workspace/projectDetection.ts
import { isForcedProjectPath, resolveTyhpProjectFile, ResolveProjectFileOptions, TYHP_PROJECT_FILE } from "../lsp/projectPath";
import { ProjectIndexSnapshot, snapshotFromOwner } from "./projectIndex";

export { TYHP_PROJECT_FILE };
export type WorkspaceSnapshot = ProjectIndexSnapshot;

export interface SnapshotPathApi {
    dirname(target: string): string;
    basename(target: string): string;
}

/**
 * Forced `tyhp.projectPath` only. Workspace scanning / include matching lives
 * in {@link ProjectIndex}.
 */
export function detectForcedProject(
    options: ResolveProjectFileOptions,
    paths: SnapshotPathApi
): WorkspaceSnapshot {
    const projectFilePath = resolveTyhpProjectFile(options);
    return snapshotFromProjectFile(projectFilePath, paths);
}

/** @deprecated Use detectForcedProject; workspace-root fallback is gone. */
export function detectWorkspaceProject(
    options: ResolveProjectFileOptions,
    paths: SnapshotPathApi
): WorkspaceSnapshot {
    return detectForcedProject(options, paths);
}

export function snapshotFromProjectFile(
    projectFilePath: string | undefined,
    paths: SnapshotPathApi
): WorkspaceSnapshot {
    if (!projectFilePath) {
        return snapshotFromOwner(undefined);
    }
    const projectDir = paths.dirname(projectFilePath);
    return {
        projectFilePath,
        projectDir,
        projectName: paths.basename(projectDir),
    };
}

export function projectStatusLabel(snapshot: WorkspaceSnapshot): string {
    const name = snapshot.projectName?.trim() ?? "";
    return name !== "" ? name : "not in a Tyhp project";
}

export function isMissingProjectLabel(label: string): boolean {
    const trimmed = label.trim();
    return trimmed === "" || trimmed === "not in a Tyhp project" || trimmed === "no project";
}

export { isForcedProjectPath };

File: tyhp-lang/phpstorm/src/main/kotlin/com/tyhp/lang/workspace/ProjectDetection.kt
package com.tyhp.lang.workspace

import com.tyhp.lang.lsp.ResolveProjectFileOptions
import com.tyhp.lang.lsp.parentDirectory
import com.tyhp.lang.lsp.resolveTyhpProjectFile

/**
 * Forced `tyhp.projectPath` snapshots. Workspace scanning / include matching
 * lives in [ProjectIndex].
 */

typealias WorkspaceSnapshot = ProjectIndexSnapshot

fun detectWorkspaceProject(options: ResolveProjectFileOptions): WorkspaceSnapshot =
    snapshotFromProjectFile(resolveTyhpProjectFile(options))

fun snapshotFromProjectFile(projectFilePath: String?): WorkspaceSnapshot {
    val file = projectFilePath?.trim()?.takeIf { it.isNotEmpty() } ?: return ProjectIndexSnapshot()
    val dir = parentDirectory(file) ?: posixDirname(file)
    val name = dir.substringAfterLast('/').substringAfterLast('\\').trim().takeIf { it.isNotEmpty() }
        ?: posixBasename(dir)
    return ProjectIndexSnapshot(
        projectFilePath = file,
        projectDir = dir,
        projectName = name,
    )
}

fun projectStatusLabel(snapshot: WorkspaceSnapshot): String {
    val name = snapshot.projectName?.trim().orEmpty()
    return name.ifEmpty { "not in a Tyhp project" }
}

/**
 * Content root that should receive `tyhp init` (and run-config cwd fallback).
 * Prefers the longest root that contains [filePath], else the first root.
 */
fun contentRootForPath(filePath: String?, contentRoots: List<String>): String? {
    val path = filePath?.trim().orEmpty()
    if (path.isNotEmpty()) {
        val match = contentRoots
            .map { it.trim() }
            .filter { it.isNotEmpty() }
            .filter { root ->
                path == root ||
                    path.startsWith("$root/") ||
                    path.startsWith("$root\\")
            }
            .maxByOrNull { it.length }
        if (match != null) {
            return match
        }
    }
    return contentRoots.firstOrNull()?.trim()?.takeIf { it.isNotEmpty() }
}
